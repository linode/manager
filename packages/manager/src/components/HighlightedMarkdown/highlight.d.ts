declare module 'highlight.js/lib/core';
declare module 'highlight.js/lib/languages/apache';
declare module 'highlight.js/lib/languages/javascript';
declare module 'highlight.js/lib/languages/yaml';
declare module 'highlight.js/lib/languages/bash';
declare module 'highlight.js/lib/languages/nginx';
