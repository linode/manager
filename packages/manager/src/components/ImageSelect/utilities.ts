import { DateTime } from 'luxon';

import { MAX_MONTHS_EOL_FILTER } from 'src/constants';

import type { ImageSelectVariant } from './ImageSelect';
import type { Image, RegionSite } from '@linode/api-v4';
import type { DisableItemOption } from 'src/components/ListItemOption';

/**
 * Given a Image Select "variant", this PR returns an
 * API filter to help you fetch the correct images.
 */
export const getAPIFilterForImageSelect = (
  variant: ImageSelectVariant | undefined
) => {
  if (variant === 'public') {
    return {
      '+order': 'desc',
      '+order_by': 'label',
      is_public: true,
    };
  }

  if (variant === 'private') {
    return { is_public: false };
  }

  return {};
};

/**
 * Unfortunately, we can't use API filtering for all of our filtering needs.
 *
 * This function exists to...
 * - filter out public Kubernetes images for the Images Select
 * - filter out images that are too far past their end-of-life date
 *
 * Please use API filtering (getAPIFilterForImageSelect) when possible!
 */
export const getFilteredImagesForImageSelect = (
  images: Image[] | undefined,
  variant: ImageSelectVariant | undefined
) => {
  if (variant === 'public') {
    // For public images, we filter out LKE images and images that are > 6 months past their `eol`
    return images?.filter(
      (image) => !image.id.includes('kube') && !isImageTooFarPastEOL(image)
    );
  }

  return images;
};

/**
 * Returns whether or not an image is too far past its end-of-life based on MAX_MONTHS_EOL_FILTER
 *
 * This function is intended to be used to filter out end-of-life images based on
 * an internal policy. See `M3-5753` for context.
 *
 * @param image an image from the API
 * @returns true if the given image is more than 6 months past its end-of-life
 */
export const isImageTooFarPastEOL = (image: Image) => {
  if (image.eol === null) {
    return false;
  }

  const imageEOL = DateTime.fromISO(image.eol);
  const now = DateTime.now();

  const differenceInMonths = now.diff(imageEOL, 'months').months;

  return differenceInMonths > MAX_MONTHS_EOL_FILTER;
};

/**
 * Returns whether or not an image is deprecated
 *
 * Unfortunately, as per `M3-5753`, we can't just simply check `deprecated` on the `Image`,
 * we must also consider the EOL date.
 */
export const isImageDeprecated = (image: Image) => {
  if (image.eol === null) {
    // If the image does not have an EOL, just use the `deprecated` field returned by the API
    return image.deprecated;
  }

  const isImageEOL = DateTime.fromISO(image.eol) < DateTime.now();

  return image.deprecated || isImageEOL;
};

interface DisabledImageOptions {
  images: Image[];
  site_type?: RegionSite;
}

/**
 * Returns images that should be disabled on the Linode Create flow.
 *
 * @returns key/value pairs for disabled images. the key is the image id and the value is why the image is disabled
 */
export const getDisabledImages = (options: DisabledImageOptions) => {
  const { images, site_type } = options;

  // Disable images that do not support distributed sites if the selected region is distributed
  if (site_type === 'distributed') {
    const disabledImages: Record<string, DisableItemOption> = {};
    for (const image of images) {
      if (!image.capabilities.includes('distributed-sites')) {
        disabledImages[image.id] = {
          reason:
            'The selected image cannot be deployed to a distributed region.',
        };
      }
    }
    return disabledImages;
  }

  return {};
};
