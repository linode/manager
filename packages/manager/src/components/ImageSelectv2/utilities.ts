import { DateTime } from 'luxon';

import { MAX_MONTHS_EOL_FILTER } from 'src/constants';

import type { ImageSelectVariant } from './ImageSelectv2';
import type { Image } from '@linode/api-v4';

/**
 * Given a Image Select "variant", this PR returns an
 * API filter to help you fetch the correct images.
 */
export const getAPIFilterForImageSelect = (
  variant: ImageSelectVariant | undefined
) => {
  if (variant === 'public') {
    return { '+order': 'desc', '+order_by': 'label', is_public: true };
  }

  if (variant === 'private') {
    return { is_public: false };
  }

  return {};
};

/**
 * Unfortunately, we can't use API filtering for all of our filtering needs.
 *
 * This function exists to...
 * - filter out public Kubernetes images for the Images Select
 * - filter out images that are too far past their end-of-life date
 *
 * Please use API filtering (getAPIFilterForImageSelect) when possible!
 */
export const getFilteredImagesForImageSelect = (
  images: Image[] | undefined,
  variant: ImageSelectVariant | undefined
) => {
  if (variant === 'public') {
    // For public images, we filter out LKE images and images that are > 6 months past their `eol`
    return images?.filter(
      (image) => !image.id.includes('kube') && !isImageTooFarPastEOL(image)
    );
  }

  return images;
};

/**
 * Returns whether or not an image is too far past its end-of-life based on MAX_MONTHS_EOL_FILTER
 *
 * This function is intended to be used to filter out end-of-life images based on
 * an internal policy. See `M3-5753` for context.
 *
 * @param image an image from the API
 * @returns true if the given image is more than 6 months past its end-of-life
 */
export const isImageTooFarPastEOL = (image: Image) => {
  if (image.eol === null) {
    return false;
  }

  const imageEOL = DateTime.fromISO(image.eol);
  const now = DateTime.now();

  const differenceInMonths = now.diff(imageEOL, 'months').months;

  return differenceInMonths > MAX_MONTHS_EOL_FILTER;
};

/**
 * Returns whether or not an image is deprecated
 *
 * Unfortunately, as per `M3-5753`, we can't just simply check `deprecated` on the `Image`,
 * we must also consider the EOL date.
 */
export const isImageDeprecated = (image: Image) => {
  if (image.eol === null) {
    // If the image does not have an EOL, just use the `deprecated` field returned by the API
    return image.deprecated;
  }

  const isImageEOL = DateTime.fromISO(image.eol) < DateTime.now();

  return image.deprecated || isImageEOL;
};
