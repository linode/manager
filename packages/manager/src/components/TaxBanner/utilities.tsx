import * as React from 'react';
import { Link } from 'react-router-dom';

export interface LinkType {
  link: string;
  text_to_replace: string;
  type: 'internal' | 'external';
}

export const generateHTMLFromString = (
  text: string,
  links: Record<string, LinkType>
) => {
  return Object.keys(links).reduce((acc, eachKey) => {
    const { text_to_replace, type, link } = links[eachKey];

    if (acc.length > 0) {
      // nth iteration after the first
      return acc.map(eachValue => {
        if (typeof eachValue === 'string') {
          return wrapStringInLink(eachValue, text_to_replace, type, link);
        } else {
          /** is a react component so no need to wrap anything in a link */
          return eachValue;
        }
      });
    } else {
      return wrapStringInLink(text, text_to_replace, type, link);
    }
  }, []);
};

export const wrapStringInLink = (
  fullString: string,
  textToReplace: string,
  type: 'internal' | 'external',
  link: string
) => {
  /**
   * split the sentence up by phrases, including the matched text in the array
   *
   * For example:
   *
   * While "I am" being the text to match, "hello world I am Marty"
   *
   * becomes
   *
   * ["hello world", "I am", "Marty"]
   *
   * See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split#Splitting_with_a_RegExp_to_include_parts_of_the_separator_in_the_result
   * for information about splitting but keeping the matched value in the array
   *
   */
  const sentenceAsArray = fullString.split(
    new RegExp(`(${textToReplace})`, 'gmi')
  );

  /**
   * iterate over the array and if we have a matched word, wrap
   * in an anchor tag or react-router link depending on the link type
   */
  const arrayWithLink = sentenceAsArray.map(eachWord => {
    if (eachWord.match(new RegExp(textToReplace, 'gmi'))) {
      return type === 'external' ? (
        <a
          key={eachWord}
          href={link}
          target="_blank"
          aria-describedby="external-site"
          rel="noopener noreferrer"
        >
          {eachWord}
        </a>
      ) : (
        <Link key={eachWord} to={link}>
          {eachWord}
        </Link>
      );
    }
    return eachWord;
  });

  /**
   * so now we have an array that looks like
   * [
   *  "hello world",
   *  "<a>I am</a>"
   *  "Marty"
   * ]
   */
  return arrayWithLink;
};
