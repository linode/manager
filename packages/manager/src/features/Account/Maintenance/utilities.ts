export const PENDING_MAINTENANCE_FILTER = Object.freeze({
  status: { '+or': ['pending', 'started'] },
});
