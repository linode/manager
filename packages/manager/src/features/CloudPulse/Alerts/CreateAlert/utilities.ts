import { omitProps } from '@linode/ui';

import type { CreateAlertDefinitionForm, MetricCriteriaForm } from './types';
import type {
  CreateAlertDefinitionPayload,
  MetricCriteria,
} from '@linode/api-v4';

// filtering out the form properties which are not part of the payload
export const filterFormValues = (
  formValues: CreateAlertDefinitionForm
): CreateAlertDefinitionPayload => {
  const values = omitProps(formValues, [
    'serviceType',
    'region',
    'engineType',
    'severity',
    'rule_criteria',
  ]);
  // severity has a need for null in the form for edge-cases, so null-checking and returning it as an appropriate type
  const severity = formValues.severity ?? 1;
  const entityIds = formValues.entity_ids;
  const rules = formValues.rule_criteria.rules;
  return {
    ...values,
    entity_ids: entityIds,
    rule_criteria: { rules: filterMetricCriteriaFormValues(rules) },
    severity,
  };
};

export const filterMetricCriteriaFormValues = (
  formValues: MetricCriteriaForm[]
): MetricCriteria[] => {
  return formValues.map((rule) => {
    const values = omitProps(rule, ['aggregation_type', 'operator', 'metric']);
    return {
      ...values,
      aggregation_type: rule.aggregation_type ?? 'avg',
      dimension_filters: rule.dimension_filters,
      metric: rule.metric ?? '',
      operator: rule.operator ?? 'eq',
    };
  });
};

export const convertToSeconds = (secondsList: string[]) => {
  return secondsList.map((second) => {
    const unit = second.slice(-1)[0];
    const number = parseInt(second.slice(0, -1), 10);
    switch (unit) {
      case 's':
        return number;
      case 'm':
        return number * 60;
      case 'h':
        return number * 3600;
      default:
        return number * 0;
    }
  });
};
