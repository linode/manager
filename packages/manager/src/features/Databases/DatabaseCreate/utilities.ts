export const determineReplicationType = (
  clusterSize: number,
  engine: string
) => {
  if (Boolean(engine.match(/mongo/))) {
    return undefined;
  }

  // If engine is a MySQL or Postgres one and it's a standalone DB instance
  if (clusterSize === 1) {
    return 'none';
  }

  // MySQL engine & cluster = semi_synch. PostgreSQL engine & cluster = asynch.
  if (Boolean(engine.match(/mysql/))) {
    return 'semi_synch';
  } else {
    return 'asynch';
  }
};

export const determineReplicationCommitType = (engine: string) => {
  // 'local' is the default.
  if (Boolean(engine.match(/postgres/))) {
    return 'local';
  }

  return undefined;
};

export const determineStorageEngine = (engine: string) => {
  // 'wiredtiger' is the default.
  if (Boolean(engine.match(/mongo/))) {
    return 'wiredtiger';
  }

  return undefined;
};

export const determineCompressionType = (engine: string) => {
  // 'none' is the default.
  if (Boolean(engine.match(/mongo/))) {
    return 'none';
  }

  return undefined;
};
