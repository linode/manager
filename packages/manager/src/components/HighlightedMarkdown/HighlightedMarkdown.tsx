import * as hljs from 'highlight.js/lib/core';
import apache from 'highlight.js/lib/languages/apache';
import bash from 'highlight.js/lib/languages/bash';
import javascript from 'highlight.js/lib/languages/javascript';
import nginx from 'highlight.js/lib/languages/nginx';
import yaml from 'highlight.js/lib/languages/yaml';
import HLJSDarkTheme from 'highlight.js/styles/a11y-dark.css?raw';
import HLJSLightTheme from 'highlight.js/styles/a11y-light.css?raw';
import * as React from 'react';

import { Typography } from 'src/components/Typography';
import 'src/formatted-text.css';
import { ThemeName } from 'src/foundations/themes';
import { unsafe_MarkdownIt } from 'src/utilities/markdown';
import { sanitizeHTML } from 'src/utilities/sanitizeHTML';
import { useColorMode } from 'src/utilities/theme';

import type { IOptions } from 'sanitize-html';

hljs.registerLanguage('apache', apache);
hljs.registerLanguage('bash', bash);
hljs.registerLanguage('javascript', javascript);
hljs.registerLanguage('nginx', nginx);
hljs.registerLanguage('yaml', yaml);

export type SupportedLanguage =
  | 'apache'
  | 'bash'
  | 'javascript'
  | 'nginx'
  | 'plaintext'
  | 'shell'
  | 'yaml';

export interface HighlightedMarkdownProps {
  className?: string;
  language?: SupportedLanguage;
  sanitizeOptions?: IOptions;
  textOrMarkdown: string;
}

export const HighlightedMarkdown = (props: HighlightedMarkdownProps) => {
  const { className, language, sanitizeOptions, textOrMarkdown } = props;
  const rootRef = React.useRef<HTMLDivElement>(null);

  const { colorMode } = useColorMode();

  /**
   * This function exists because we use Hightlight.js and it does not have a built-in
   * way to programaticly change the theme.
   *
   * We must manually switch our Highlight.js theme's CSS when our theme is changed.
   */
  const handleThemeChange = async (theme: ThemeName) => {
    const THEME_STYLE_ID = 'hljs-theme';
    const existingStyleTag = document.getElementById(THEME_STYLE_ID);

    if (existingStyleTag) {
      // If the style tag already exists in the <head>, just update the css content.
      existingStyleTag.innerHTML =
        theme === 'light' ? HLJSLightTheme : HLJSDarkTheme;
    } else {
      // The page has been loaded and we need to manually append our Hightlight.js
      // css so we can easily change it later on.
      const styleTag = document.createElement('style');
      styleTag.id = THEME_STYLE_ID;
      styleTag.innerHTML = theme === 'light' ? HLJSLightTheme : HLJSDarkTheme;
      document.head.appendChild(styleTag);
    }
  };

  React.useEffect(() => {
    handleThemeChange(colorMode);
  }, [colorMode]);

  /**
   * If the language prop is provided, we'll try to override the language
   * auto detection to specify the selected language.
   */
  React.useEffect(() => {
    if (language) {
      hljs.configure({
        languages: [language],
      });
    }
  }, [language]);

  const unsafe_parsedMarkdown = unsafe_MarkdownIt.render(textOrMarkdown);

  const sanitizedHtml = sanitizeHTML({
    // eslint-disable-next-line xss/no-mixed-html
    options: sanitizeOptions,
    sanitizingTier: 'flexible',
    text: unsafe_parsedMarkdown,
  });

  // Adapted from https://stackblitz.com/edit/react-highlighted-markdown?file=highlighted-markdown.tsx
  // All the safety checking is due to a reported error from certain versions of FireFox.
  React.useEffect(() => {
    try {
      if (rootRef.current) {
        const blocks = rootRef.current.querySelectorAll('pre code') ?? [];
        const len = blocks.length ?? 0;
        let i = 0;
        for (i; i < len; i++) {
          hljs.highlightBlock(blocks[i]);
        }
      }
    } catch {
      // do nothing, it's not the end of the world if we can't highlight Markdown.
    }
  }, [textOrMarkdown]);

  return (
    <Typography
      dangerouslySetInnerHTML={{
        __html: sanitizedHtml,
      }}
      sx={(theme) => ({
        '& .hljs': {
          color: theme.color.offBlack,
        },
      })}
      className={`formatted-text ${className}`}
      ref={rootRef}
    />
  );
};
