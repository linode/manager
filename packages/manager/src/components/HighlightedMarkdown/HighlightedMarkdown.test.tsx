import * as React from 'react';

import { renderWithTheme } from 'src/utilities/testHelpers';
import HighlightedMarkdown from './HighlightedMarkdown';

const sampleMarkdown =
  '# Some markdown \n ```\n const x = function() { return true; }\n```';

describe('HighlightedMarkdown component', () => {
  it('should highlight text consistently', () => {
    const { asFragment } = renderWithTheme(
      <HighlightedMarkdown textOrMarkdown={sampleMarkdown} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
