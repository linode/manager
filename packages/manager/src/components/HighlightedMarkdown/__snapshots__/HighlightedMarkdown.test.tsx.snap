// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightedMarkdown component should highlight text consistently 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root makeStyles-root-1 formatted-text MuiTypography-body1"
  />
  <h1>
    Some markdown
  </h1>
  

  <pre>
    <code
      class="hljs javascript"
    >
      <span
        class="hljs-keyword"
      >
        const
      </span>
       x = 
      <span
        class="hljs-function"
      >
        <span
          class="hljs-keyword"
        >
          function
        </span>
        (
        <span
          class="hljs-params"
        />
        ) 
      </span>
      { 
      <span
        class="hljs-keyword"
      >
        return
      </span>
       
      <span
        class="hljs-literal"
      >
        true
      </span>
      ; }

    </code>
  </pre>
  <p />
</DocumentFragment>
`;
